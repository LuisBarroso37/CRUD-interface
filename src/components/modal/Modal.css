.modal-background {
  position: fixed;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 50;
  background: rgba(255, 255, 255, 0.8);
  display: flex;
  justify-content: center;
  align-items: center;
}

.modal {
  width: 500px;
  background: #ffffff;
  box-shadow: 0px 4px 12px rgba(0, 0, 0, 0.12);
  border-radius: 24px;
  padding: 0px 20px 20px 20px;
}

.modal-header {
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.modal-title {
  height: 28px;
  font-size: 24px;
  line-height: 28px;
}

.modal-close {
  color: #bfbfbf;
  font-size: 25px;
  cursor: pointer;
}

.modal-close:hover {
  color: #a6a6a6;
}
