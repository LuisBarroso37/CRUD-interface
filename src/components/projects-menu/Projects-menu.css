.projects-menu {
  width: 442px;
  background: #ffffff;
  box-shadow: 0px 4px 12px rgba(0, 0, 0, 0.12);
  border-radius: 24px;
  padding-bottom: 10px;
}

.menu-header {
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 0 25px;
}

.projects-title {
  width: 89px;
  height: 28px;
  font-size: 24px;
  line-height: 28px;
}

.add-project-button {
  color: rgba(0, 0, 0, 0.8);
  font-size: 2rem;
  border: none;
  background: #ffffff;
  cursor: pointer;
}

.project {
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 0 25px;
}

.project-info {
  display: flex;
  flex-direction: column;
}

.project-name {
  font-size: 18px;
  line-height: 21px;
  margin-bottom: 0;
}

.project-users {
  font-weight: normal;
  font-size: 14px;
  line-height: 16px;
}

.icons {
  display: flex;
  justify-content: space-evenly;
  width: 25%;
}

.crud-icon {
  color: rgba(0, 0, 0, 0.64);
  width: 100%;
  cursor: pointer;
}
