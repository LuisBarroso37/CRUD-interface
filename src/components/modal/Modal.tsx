import React, { useRef, useCallback, useEffect } from 'react';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';

import './Modal.css';

import ProjectModal from '../project-modal/Project-modal';
import UsersModal from '../users-modal/Users-modal';
import { IRootReducerState } from '../../redux/root-reducer';
import {
  IModalActionsDispatch,
  setModal,
  setUseModal,
} from '../../redux/modal/modal.actions';

type IModalProps = ReturnType<typeof mapDispatchToProps> &
  ReturnType<typeof mapStateToProps>;

const Modal: React.FC<IModalProps> = ({
  setUseModal,
  setModal,
  projects,
  num,
  buttonLabel,
}: IModalProps) => {
  const ref = useRef(null);

  // Handler function
  const closeModal = useCallback(() => {
    setUseModal();
    setModal('');
  }, [setModal, setUseModal]);

  // Close Modal when click is performed outside of modal
  const useOnClickOutside = (
    ref: any,
    handler: (event: MouseEvent | TouchEvent) => void
  ) => {
    useEffect(() => {
      const listener = (event: MouseEvent | TouchEvent) => {
        // Do nothing if clicking ref's element or descendent elements
        if (!ref.current || ref.current.contains(event.target)) {
          return;
        }

        handler(event);
      };

      // Add the listeners on component mount
      document.addEventListener('mousedown', listener);
      document.addEventListener('touchstart', listener);

      // Remove the listeners on component unmount
      return () => {
        document.removeEventListener('mousedown', listener);
        document.removeEventListener('touchstart', listener);
      };
    }, [ref, handler]);
  };

  useOnClickOutside(ref, () => closeModal());

  return (
    <div className='modal-background'>
      <div ref={ref} className='modal'>
        <div className='modal-header'>
          <p className='modal-title'>
            {buttonLabel === 'Invite new users'
              ? 'Users settings'
              : buttonLabel === 'Edit project'
              ? `Edit "${projects[num].name}"`
              : 'Add project'}
          </p>
          <i
            className='fas fa-times modal-close'
            onClick={() => closeModal()}
          ></i>
        </div>
        {buttonLabel === 'Invite new user' ? <UsersModal /> : <ProjectModal />}
      </div>
    </div>
  );
};

const mapDispatchToProps = (dispatch: Dispatch<IModalActionsDispatch>) => ({
  setUseModal: () => dispatch(setUseModal()),
  setModal: (action: string) => dispatch(setModal(action)),
});

const mapStateToProps = (state: IRootReducerState) => ({
  modal: state.modal.modalAction,
  useModal: state.modal.useModal,
  num: state.projects.numProject,
  projects: state.projects.projectsDB,
  buttonLabel: state.projects.buttonLabel,
});

export default connect(mapStateToProps, mapDispatchToProps)(Modal);
