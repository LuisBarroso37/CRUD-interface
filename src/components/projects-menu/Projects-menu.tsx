import React from 'react';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';

import './Projects-menu.css';

import Modal from '../modal/Modal';
import {
  IRootReducerState,
  IRootReducerActions,
} from '../../redux/root-reducer';
import { setModal, setUseModal } from '../../redux/modal/modal.actions';
import {
  setProjects,
  IProjects,
  setButtonLabel,
  setNumProject,
} from '../../redux/projects/projects.actions';

type IProjectsMenuProps = ReturnType<typeof mapDispatchToProps> &
  ReturnType<typeof mapStateToProps>;

export const ProjectsMenu: React.FC<IProjectsMenuProps> = ({
  setModal,
  useModal,
  setUseModal,
  projects,
  setProjects,
  setNumProject,
  setButtonLabel,
}) => (
  <div className='projects-menu'>
    <div className='menu-header'>
      <p className='projects-title'>Projects</p>
      <button
        className='add-project-button'
        onClick={() => {
          setUseModal();
          setModal('add-project');
          setNumProject(projects.length + 1);
          setButtonLabel('Add project');
        }}
      >
        +
      </button>
    </div>
    
    <div className='projects-content'>
      {projects.map((project, i) => (
        <div key={i} className='project'>
          <div className='project-info'>
            <p className='project-name'>{project.name}</p>
            <p className='project-users'>{project.collaborators} collaborators</p>
          </div>
          <div className='icons'>
            <i
              className='fas fa-user-plus crud-icon'
              onClick={() => {
                setUseModal();
                setModal('add-user');
                setNumProject(i);
                setButtonLabel('Invite new user');
              }}
            ></i>
            <i
              className='far fa-edit crud-icon'
              onClick={() => {
                setUseModal();
                setModal('edit-project');
                setNumProject(i);
                setButtonLabel('Edit project');
              }}
            ></i>
            <i
              className='far fa-trash-alt crud-icon'
              onClick={() => {
                let projectsCopy = [...projects];
                projectsCopy.splice(i, 1);
                setProjects(projectsCopy);
              }}
            ></i>
          </div>
        </div>
      ))}
    </div>

    {useModal ? <Modal /> : null}
  </div>
);

const mapDispatchToProps = (dispatch: Dispatch<IRootReducerActions>) => ({
  setUseModal: () => dispatch(setUseModal()),
  setModal: (action: string) => dispatch(setModal(action)),
  setNumProject: (index: number) => dispatch(setNumProject(index)),
  setProjects: (projects: IProjects) => dispatch(setProjects(projects)),
  setButtonLabel: (label: string) => dispatch(setButtonLabel(label)),
});

const mapStateToProps = (state: IRootReducerState) => ({
  useModal: state.modal.useModal,
  projects: state.projects.projectsDB,
});

export default connect(mapStateToProps, mapDispatchToProps)(ProjectsMenu);
