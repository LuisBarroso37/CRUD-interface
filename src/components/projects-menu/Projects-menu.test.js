import { shallow } from 'enzyme';

import { ProjectsMenu } from './Projects-menu';
import PROJECTS from './PROJECTS_DATA';
import Modal from '../modal/Modal';

describe('Projects Menu component testing', () => {
  let wrapper;
  let mockProps;
  let setModal;
  let setUseModal;
  let setProjects;
  let setNumProject;
  let setButtonLabel;

  beforeEach(() => {
    setModal = jest.fn();
    setUseModal = jest.fn();
    setProjects = jest.fn((newProjects) => (mockProps.projects = newProjects));
    setNumProject = jest.fn();
    setButtonLabel = jest.fn();

    mockProps = {
      projects: PROJECTS,
      setModal,
      useModal: '',
      setUseModal,
      setProjects,
      setNumProject,
      setButtonLabel,
    };

    wrapper = shallow(<ProjectsMenu {...mockProps} />);
  });

  it('expect to render ProjectMenu component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('expect actions to be dispatched when add-project-icon button is clicked', () => {
    wrapper.find('.add-project-button').simulate('click');
    expect(setUseModal).toHaveBeenCalled();
    expect(setModal).toHaveBeenCalled();
    expect(setNumProject).toHaveBeenCalled();
    expect(setButtonLabel).toHaveBeenCalled();
  });

  it('expect actions to be dispatched when add-user icon is clicked', () => {
    wrapper.find('.fa-user-plus').first().simulate('click');
    expect(setUseModal).toHaveBeenCalled();
    expect(setModal).toHaveBeenCalled();
    expect(setNumProject).toHaveBeenCalled();
    expect(setButtonLabel).toHaveBeenCalled();
  });

  it('expect actions to be dispatched when edit-project icon is clicked', () => {
    wrapper.find('.fa-edit').first().simulate('click');
    expect(setUseModal).toHaveBeenCalled();
    expect(setModal).toHaveBeenCalled();
    expect(setNumProject).toHaveBeenCalled();
    expect(setButtonLabel).toHaveBeenCalled();
  });

  it('expect modal to be rendered when add-user icon is clicked', () => {
    const newMockProps = {
      projects: PROJECTS,
      setModal,
      useModal: 'add-user',
      setUseModal,
      setProjects,
      setNumProject,
      setButtonLabel,
    };

    const newWrapper = shallow(<ProjectsMenu {...newMockProps} />);

    const modal = newWrapper.find(Modal);
    expect(modal).toHaveLength(1);
  });

  it('expect actions to be dispatched when delete-project icon is clicked', () => {
    const newProjects = [...mockProps.projects];
    newProjects.shift(); // Remove first project

    wrapper.find('.fa-trash-alt').first().simulate('click');
    expect(setProjects).toHaveBeenCalled();
    expect(setProjects(newProjects)).toEqual(mockProps.projects);
  });
});
