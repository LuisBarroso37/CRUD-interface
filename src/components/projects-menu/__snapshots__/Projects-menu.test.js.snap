// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Projects Menu component testing expect to render ProjectMenu component 1`] = `ShallowWrapper {}`;
